modules/core/src/smithy4s/Blob.scala

```diff
@@ -18,101 +18,238 @@ package smithy4s
 
 import java.nio.ByteBuffer
 import java.util.Base64
+import scala.collection.immutable.Queue
 import java.nio.charset.StandardCharsets
-
+import java.io.OutputStream
+
+// scalafmt: {maxColumn = 120}
+/**
+  * A Blob represents an arbitrary piece of binary data that fits in memory.
+  *
+  * Its underlying data structure enables several types of layouts, as well as efficient concatenation.
+  */
 sealed trait Blob {
-  def toArray: Array[Byte]
-  def buffer: ByteBuffer
-  def isEmpty: Boolean
+
+  def apply(i: Int): Byte
   def size: Int
-  def sameBytesAs(other: Blob): Boolean
-  final def toArrayBlob: Blob = Blob(toArray)
+  def isEmpty: Boolean
+
+  def foreach(f: Byte => Unit) = {
+    var i = 0
+    while (i < size) { f(apply(i)); i += 1 }
+  }
+
+  def foreachWithIndex(f: (Byte, Int) => Unit) = {
+    var i = 0
+    while (i < size) { f(apply(i), i); i += 1 }
+  }
+
+  def toArraySliceBlob: Blob.ArraySliceBlob =
+    new Blob.ArraySliceBlob(toArrayUnsafe, 0, size)
+
+  def toArray: Array[Byte] = {
+    val result = Array.ofDim[Byte](size)
+    foreachWithIndex((b, i) => result(i) = b)
+    result
+  }
+
+  def toArrayUnsafe: Array[Byte] = toArray
+
+  def sameBytesAs(other: Blob): Boolean = {
+    size == other.size && {
+      var i = 0
+      var result = true
+      while (i < size && result) {
+        result = this(i) == other(i)
+        i += 1
+      }
+      result
+    }
+  }
+
+  def asByteBuffer(offset: Int, size: Int): ByteBuffer = {
+    val arr = new Array[Byte](size)
+    copyToArray(arr, 0, offset, size)
+    ByteBuffer.wrap(arr)
+  }
+  def asByteBuffer: ByteBuffer = asByteBuffer(0, size)
+  def asByteBufferUnsafe(offset: Int, size: Int): ByteBuffer = asByteBuffer(offset, size)
+  def asByteBufferUnsafe: ByteBuffer = asByteBuffer(0, size)
```

Comment:
what are the `unsafe` alternative for? they seem to have that same underlying implementation

Reply:
`asByteBuffer` may return a read-only `ByteBuffer` to protect the backing memory. `asByteBufferUnsafe` unsafely exposes the backing memory. This makes a meaningful difference if you want to obtain the `Array[Byte]` backing a `ByteBuffer`.